File: solutions/0257-binary-tree-paths/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        std::vector<std::string> v_ret;
        if (!root) return v_ret;
        if (!root->left && !root->right) {
            v_ret.push_back(std::to_string(root->val));
            return v_ret;
        }
        postorderHelper(root, v_ret);
        return v_ret;
    }
private:
    void postorderHelper(TreeNode* root, std::vector<std::string>& paths) {
        std::stack<TreeNode*> s;
        std::stack<std::string> p_s;
        s.push(root);
        p_s.push(std::to_string(root->val));
        while (!s.empty()) {
            root = s.top();
            s.pop();
            std::string r_path(p_s.top());
            p_s.pop();
            // Left - first
            if (root->left) {
                s.push(root->left);
                p_s.push(r_path + "->" + std::to_string(root->left->val));
            }
            // Right next
            if (root->right) {
                s.push(root->right);
                p_s.push(r_path + "->" + std::to_string(root->right->val));
            }
            // Process Node
            // leaf node check
            if (!root->left && !root->right) {
                // std::cout << r_path << std::endl;
                paths.push_back(r_path);
            }
        }
    }

};

File: solutions/0257-binary-tree-paths/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import lls "github.com/emirpasic/gods/stacks/linkedliststack"

func binaryTreePaths(root *TreeNode) []string {
    var res []string
    if root == nil { return res }
    if root.Left == nil && root.Right == nil { return []string{strconv.Itoa(root.Val)} }
    var postOrderHelper func(*TreeNode)
    postOrderHelper = func(root *TreeNode) {
        // var ssb []strings.Builder
        var sb strings.Builder
        sb.WriteString(strconv.Itoa(root.Val))
        s := lls.New()
        ps := lls.New()
        s.Push(root)
        ps.Push(&sb)
        for !s.Empty() {
            curEle, _ := s.Pop()
            cur := curEle.(*TreeNode)
            psEle, _ := ps.Pop()
            var psb strings.Builder
            psb.WriteString(psEle.(*strings.Builder).String())
            if cur.Left != nil {
                s.Push(cur.Left)
                var psb strings.Builder
                psb.WriteString(psEle.(*strings.Builder).String())
                psb.WriteString("->")
                psb.WriteString(strconv.Itoa(cur.Left.Val))
                ps.Push(&psb)
            }
            if cur.Right != nil {
                s.Push(cur.Right)
                var psb strings.Builder
                psb.WriteString(psEle.(*strings.Builder).String())
                psb.WriteString("->")
                psb.WriteString(strconv.Itoa(cur.Right.Val))
                ps.Push(&psb)
            }

            if cur.Left == nil && cur.Right == nil {
                res = append(res, psb.String())
                // fmt.Println(len(res))
                // fmt.Println(res)
                // fmt.Println("*******")
            }
        }
    }
    postOrderHelper(root)
    return res
}
